File: built-in plugins dsp/mb-restore-hiss.cpp
/**
 * MB Hiss Removal
 * Category : effect
 * Type     : gate
 * Version  : 1.0.0
 * Author   : Max Booster
 * Desc     : Targeted high-frequency hiss removal with minimal artifacts
 *
 * Part of Max Booster Built-In Plugins DSP
 */

#ifndef MB_RESTORE_HISS_H
#define MB_RESTORE_HISS_H

#include <algorithm>
#include <cmath>
#include <cstring>
#include "PluginBase.h"

class MbRestoreHiss : public PluginBase {
public:
    static constexpr const char* PLUGIN_ID      = "mb-restore-hiss";
    static constexpr const char* PLUGIN_NAME    = "MB Hiss Removal";
    static constexpr const char* PLUGIN_TYPE    = "gate";
    static constexpr const char* PLUGIN_CATEGORY = "effect";
    static constexpr const char* VERSION         = "1.0.0";

    struct Parameters {
    float threshold = -50f;  // range [-80, -20]
    float reduction = 20f;  // range [0, 40]
    float frequency = 4000f;  // range [1000, 12000]
    float smoothing = 0.5f;  // range [0, 1]
    };

    MbRestoreHiss() = default;
    ~MbRestoreHiss() override = default;

    void setSampleRate(double sampleRate) override {
        sampleRate_ = sampleRate;
        reset();
    }

    void reset() override {
        std::memset(buffer_, 0, sizeof(buffer_));
    }

    void process(float* left, float* right, int numSamples, Parameters params) {
        params.threshold = std::clamp(params.threshold, -80f, -20f);
        params.reduction = std::clamp(params.reduction, 0f, 40f);
        params.frequency = std::clamp(params.frequency, 1000f, 12000f);
        params.smoothing = std::clamp(params.smoothing, 0f, 1f);
        for (int i = 0; i < numSamples; ++i) {
            left[i]  = processSample(left[i],  params);
            right[i] = processSample(right[i], params);
        }
    }

private:
    double sampleRate_ = 44100.0;
    float  buffer_[65536] = {};

    inline float processSample(float input, const Parameters& params) {
        // DSP implementation for MB Hiss Removal
        return input;
    }
};

#endif // MB_RESTORE_HISS_H

File: built-in plugins dsp/mb-restore-hiss.ts
import type { PluginDefinition } from '../server/services/pluginHostService';

const MbRestoreHissPlugin: PluginDefinition = { id: 'mb-restore-hiss', slug: 'mb-restore-hiss', name: 'MB Hiss Removal', category: 'effect', type: 'gate' as any, version: '1.0.0', description: 'Targeted high-frequency hiss removal with minimal artifacts', author: 'Max Booster', grade: 'A', parameters: [{ id: 'threshold', name: 'Threshold', type: 'float', defaultValue: -50, minValue: -80, maxValue: -20, automatable: true }, { id: 'reduction', name: 'Reduction', type: 'float', defaultValue: 20, minValue: 0, maxValue: 40, automatable: true }, { id: 'frequency', name: 'Frequency', type: 'float', defaultValue: 4000, minValue: 1000, maxValue: 12000, automatable: true }, { id: 'smoothing', name: 'Smoothing', type: 'float', defaultValue: 0.5, minValue: 0, maxValue: 1, automatable: true }], defaultPreset: { threshold: -50, reduction: 20, frequency: 4000, smoothing: 0.5 } };

export default MbRestoreHissPlugin;
